// Copyright 2025 John Manferdelli, All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// or in the the file LICENSE-2.0.txt in the top level sourcedirectory
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License
// File:acl.proto

syntax="proto2";

// These are existing certifier protos which may be useful

message binary_blob_message {
  optional string blob                          = 1;
};

message encrypted_message {
  optional string encryption_identifier         = 1;
  optional string message_identifier            = 2;
  optional string source                        = 3;
  optional string destination                   = 4;
  optional string date                          = 5;
  optional bytes buffer                         = 6;
};

message buffer_list {
  repeated bytes blobs                          = 1;
};

message signature_message {
  optional string encryption_algorithm_name     = 1;
  optional string key_name                      = 2;
  optional bytes signature                      = 3;
  optional string signer_name                   = 4;
};

message rsa_message {
  optional bytes public_modulus             = 1;
  optional bytes public_exponent            = 2;
  optional bytes private_exponent           = 3;
  optional bytes private_p                  = 4;
  optional bytes private_q                  = 5;
  optional bytes private_dp                 = 6;
  optional bytes private_dq                 = 7;
  optional bytes private_iqmp               = 8;
};

message point_message {
 optional bytes x                           = 1;
 optional bytes y                           = 2;
};

message ecc_message {
  optional string curve_name                = 1;
  optional bytes curve_p                    = 2;
  optional bytes curve_a                    = 3;
  optional bytes curve_b                    = 4;
  optional point_message base_point         = 5;
  optional point_message public_point       = 6;
  optional bytes order_of_base_point        = 7;
  optional bytes private_multiplier         = 8;
};

// Key types: "rsa-2048-public" "ecc-384-public",
//    "rsa-2048-private" "ecc-384-private",
//    "aes-256", "aes-256-hmac-sha-256", etc.
// Principal formats: "vse-entity"
// Key formats: "vse-key", "x509-cert"
message key_message {
  optional string key_name                  = 1;
  optional string key_type                  = 2;
  optional string key_format                = 3;
  optional rsa_message rsa_key              = 4;
  optional ecc_message ecc_key              = 5;
  optional bytes secret_key_bits            = 6;
  optional bytes certificate                = 7;
  optional bytes other_key_formats          = 8;
  optional string not_before                = 9;
  optional string not_after                 = 10;
};

// These are the new acl specific protos

message principal_message {
  optional string principal_name                      = 1;
  optional string authentication_algorithm            = 2;
  optional bytes credential                           = 3;
};

message audit_info {
  optional string audit_file                          = 1;
  optional string protocol                            = 2;
};

// for now every resource is in its own file named by resource_location
// file is encrypted with resource_key
// if audit is required, logging info is in audit info
message resource_message {
  optional string resource_identifier                 = 1;
  optional string resource_location                   = 2;
  optional key_message resource_key                   = 3;
  optional bool audit_required                        = 4;
  optional audit_info log                             = 5;
  optional string time_created                        = 6;
  optional string time_last_written                   = 7;
  repeated string readers                             = 8;
  repeated string writers                             = 9;
  repeated string deleters                            =10;
  repeated string creators                            =11;
};

message resource_list {
  repeated resource_message resources                 = 1;
};

message principal_list {
  repeated principal_message principals               = 1;
};

message access_list {
  optional string resource                            = 1;
  repeated string prins                               = 2;
};


